src/backend/distributed/planner/shard_pruning.c

```diff
@@ -1248,13 +1257,38 @@ TransformPartitionRestrictionValue(Var *partitionColumn, Const *restrictionValue
 	/* if still not a constant, no immutable coercion matched */
 	if (!IsA(transformedValue, Const))
 	{
+		if (!missingOk)
+		{
+			ErrorTypesDontMatch(partitionColumn->vartype, partitionColumn->varcollid,
+								restrictionValue->consttype,
+								restrictionValue->constcollid);
+		}
+
 		return NULL;
 	}
 
 	return (Const *) transformedValue;
 }
 
 
+/*
+ * ErrorTypesDontMatch throws an error explicitly printing the type names.
+ */
+static void
+ErrorTypesDontMatch(Oid firstType, Oid firstCollId, Oid secondType, Oid secondCollId)
+{
+	Datum firstTypename =
+		DirectFunctionCall1Coll(regtypeout, firstCollId, ObjectIdGetDatum(firstType));
+
+	Datum secondTypename =
+		DirectFunctionCall1Coll(regtypeout, secondCollId, ObjectIdGetDatum(secondType));
+
+	ereport(ERROR, (errmsg("Cannot coerce %s to %s",
```

Comment:
is there a way to test this?

Reply:
Probably not because Postgres doesn't allow to pass types that are not coercible to each other. But, this felt  more  secure in case  we miss anything. 

(I can add a UDF to test, though I  tested manually by calling this function in the code  and checking the output)